packages/validator/src/processor.ts

```diff
@@ -0,0 +1,204 @@
+import * as utils from './utils/utils';
+import * as logger from './utils/logger';
+import { validateCondition } from './validators/conditionValidator';
+import { validateRegexp } from './validators/regexpValidator';
+import { validateOptional } from './validators/optionalValidator';
+import { isAnyParamValid } from './validators/anyValidator';
+import { Log, Result, Roots } from './types';
+import { execute } from './utils/action';
+import fs from 'fs';
+
+const apiTemplate = JSON.parse(fs.readFileSync('templates/apiSpecifications.json', 'utf8'));
+const oisTemplate = JSON.parse(fs.readFileSync('templates/ois.json', 'utf8'));
+const endpointsTemplate = JSON.parse(fs.readFileSync('templates/endpoints.json', 'utf8'));
+
+/**
+ * Recursion validating provided specification against template
+ * @param specs - specification that is being validated
+ * @param template - template the specification is validated against
+ * @param paramPath - string of parameters separated by ".", representing path to current specs location (empty string in root)
+ * @param nonRedundantParams - object containing all required and optional parameters that are being used
```

Comment:
I'm having trouble understanding what `nonRedundantParams` is used for. My best guess: this is used to keep track of extra fields (and nothing else), correct?

If so, I think it adds too much complexity to pass around `nonRedundantParams` between practically all subroutines.

What I would do to check for extra fields (and missing params) is: _before_ the for-loop in `processSpecs` do something like this:
```
let specKeys = Object.keys(specs);
// filter out "special" keys from template; i.e. those starting with "__"
let templateKeys = Object.keys(template).filter(key => !key.startsWith('__'));
 
// convert Arrays to Sets for easier set operations
specKeys = new Set(specKeys);
templateKeys = new Set(templateKeys);
 
// check for missing params
let keysNotInSpec = new Set([...templateKeys].filter(x => !specKeys.has(x)));
if (keysNotInSpec.size > 0) {
    messages.push(...) // push missing param error message
}
 
// check for extra fields
if (!Object.keys(template).includes("__ignore")) {
    let keysNotinTemplate = new Set([...specKeys].filter(x => !TemplateKeys.has(x)));
    if (keysNotInTemplate.size > 0) {
        messages.push(...) // push extra fields error message
    }
}
```

This way, you check for extra fields and missing specs in a more explicit and isolated way. (Apologies for any potential syntax errors; I rarely use js/ts.)

Reply:
Also: if there are any missing params and/or extra fields, I think it's safe to just return at this point with an error message.

Reply:
Yes this seems simple enough, however it's not going to cover way too many cases. Im aware that passing around "useless" variable is less than ideal, but it truly is the best solution that i came up with. Back to your solution, if i understand it correctly there are 2 fairly big issues right on top of my head (its going to be probably more though). First one should be what about `__keyRegexp`? Some parameters are not extra even though they are not explicitly given in template. Another one would be `__conditions`. Parameters that are not in template but used in valid `__condition` are also not extra, keep in mind `__rootThen` is evaluated all the way from root, which complicates things even more.
Im opened to any suggestions about this, as im not happy with my solution that much, but i dont really see any easy way out of this.

Reply:
Yea, you're right. I didn't think about `__rootThen`; that really complicates things since I thought you could do missing/extra param check at each "level", but `__rootThen` wouldn't allow that. Okay, leave it as-is.